python/cudf/cudf/core/join/_join_helpers.py

```diff
@@ -0,0 +1,194 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.
+from __future__ import annotations
+
+import warnings
+from typing import TYPE_CHECKING, Any, Iterable
+
+import numpy as np
+import pandas as pd
+
+import cudf
+from cudf.core.dtypes import CategoricalDtype
+
+if TYPE_CHECKING:
+    from cudf._typing import Dtype
+    from cudf.core.column import ColumnBase
+    from cudf.core.frame import Frame
+
+
+class _Indexer:
+    # Indexer into a column (either a data column or index level).
+    #
+    # >>> df
+    #    a
+    # b
+    # 4  1
+    # 5  2
+    # 6  3
+    # >>> _Indexer("a", column=True).get(df)  # returns column "a" of df
+    # >>> _Indexer("b", index=True).get(df)  # returns index level "b" of df
+
+    def __init__(self, name: Any, column=False, index=False):
+        self.name = name
+        self.column, self.index = column, index
+
+    def get(self, obj: Frame) -> ColumnBase:
+        # get the column from `obj`
+        if self.column:
+            return obj._data[self.name]
+        else:
+            if obj._index is not None:
+                return obj._index._data[self.name]
+        raise KeyError()
+
+    def set(self, obj: Frame, value: ColumnBase):
+        # set the colum in `obj`
+        if self.column:
+            obj._data[self.name] = value
+        else:
+            if obj._index is not None:
+                obj._index._data[self.name] = value
+            else:
+                raise KeyError()
+
+
+def _frame_select_by_indexers(
+    frame: Frame, indexers: Iterable[_Indexer]
+) -> Frame:
+    # Select columns from the given `Frame` using `indexers`,
+    # and return a new `Frame`.
+    index_data = frame._data.__class__()
+    data = frame._data.__class__()
+
+    for idx in indexers:
+        if idx.index:
+            index_data[idx.name] = idx.get(frame)
+        else:
+            data[idx.name] = idx.get(frame)
+
+    result_index = cudf.Index._from_data(index_data) if index_data else None
+    result = cudf.core.frame.Frame(data=data, index=result_index)
+    return result
+
+
+def _match_join_keys(lcol: ColumnBase, rcol: ColumnBase, how: str) -> Dtype:
+    # cast the keys lcol and rcol to a common dtype
+
+    ltype = lcol.dtype
+    rtype = rcol.dtype
+
+    # if either side is categorical, different logic
+    if isinstance(ltype, CategoricalDtype) or isinstance(
+        rtype, CategoricalDtype
+    ):
+        return _match_categorical_dtypes(ltype, rtype, how)
+
+    if pd.api.types.is_dtype_equal(ltype, rtype):
+        return ltype
+
+    if (np.issubdtype(ltype, np.number)) and (np.issubdtype(rtype, np.number)):
```

Comment:
IIRC this handled upcasting numeric dtypes to string previously, does that still work? What happens if we try and merge `[1,2,3]` to `['1','2','3']` now?

Reply:
It previously did, but as I discovered, in a way that depends on a chain of implicit behaviours :D 

* Currently, the [`_input_to_llibcudf_castrulles`](https://github.com/rapidsai/cudf/blob/branch-0.19/python/cudf/cudf/core/join/casting_logic.py#L102) function returns `None` in the case you mentioned

* We pass that `None` to `astype()` which happens to default to callling [`as_numerical_column`](https://github.com/rapidsai/cudf/blob/branch-0.18/python/cudf/cudf/core/column/column.py#L1009)

* In `as_numerical_column`, `dtype=None` happens to result in a `float64`:

```python
In [2]: np.dtype(None)
Out[2]: dtype('float64')
```

---

Sorry to answer your question with more questions:

* Do we really want to support this behaviour or force users to do the cast themselves?
* If so, what should the behaviour be when joining a string column with a numeric (or datetime, or decimal, or...) column?

If we do decide to support this, we can take this opportunity to make the behaviour more explicit here.

Reply:
I would have naively thought that pandas promoted to `object` here. But it appears it does not:

```
ValueError: You are trying to merge on int64 and object columns. If you wish to proceed you should use pd.concat
```
Originally I had thought this would just fall into `find_common_type` and give us back `object`. But it's simpler to just follow pandas in this case and throw.